From 78b5302d73c5d61736e86c035a8436da30f27693 Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:26:21 -0400
Subject: [md] raid10: fix memory leak when reshaping a RAID10

Message-id: <1409563584-9000-80-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 90015
O-Subject: [RHEL7.1 PATCH 79/82] md/raid10: fix memory leak when reshaping a RAID10.
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: NeilBrown <neilb@suse.de>

raid10 reshape clears unwanted bits from a bio->bi_flags using
a method which, while clumsy, worked until 3.10 when BIO_OWNS_VEC
was added.
Since then it clears that bit but shouldn't.  This results in a
memory leak.

So change to used the approved method of clearing unwanted bits.

As this causes a memory leak which can consume all of memory
the fix is suitable for -stable.

Fixes: a38352e0ac02dbbd4fa464dc22d1352b5fbd06fd
Cc: stable@vger.kernel.org (v3.10+)
Reported-by: mdraid.pkoch@dfgh.net (Peter Koch)
Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit ce0b0a46955d1bb389684a2605dbcaa990ba0154)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/raid10.c b/drivers/md/raid10.c
index b7f6134..68e8b3f 100644
--- a/drivers/md/raid10.c
+++ b/drivers/md/raid10.c
@@ -4422,7 +4422,7 @@ read_more:
  read_bio->bi_private = r10_bio;
  read_bio->bi_end_io = end_sync_read;
  read_bio->bi_rw = READ;
- read_bio->bi_flags &= ~(BIO_POOL_MASK - 1);
+ read_bio->bi_flags &= (~0UL << BIO_RESET_BITS);
  read_bio->bi_flags |= 1 << BIO_UPTODATE;
  read_bio->bi_vcnt = 0;
  read_bio->bi_size = 0;
-- 
1.7.1